File: 1364-tuple-with-same-product/1364-tuple-with-same-product.java
import java.util.HashMap;

class Solution {
    public int tupleSameProduct(int[] arr) {
        int count = 0;
        HashMap<Integer, Integer> f = new HashMap<>();

        // Count frequency of products
        for (int i = 0; i < arr.length; i++) {
            for (int j = i + 1; j < arr.length; j++) { 
                int x = arr[i] * arr[j];
                f.put(x, f.getOrDefault(x, 0) + 1);
            }
        }

       
        for (Integer key : f.keySet()) {
            int freq = f.get(key);
            if (freq > 1) {
                count += (8 * (freq * (freq - 1)) / 2); //  tuple count formula
            }
        }
        return count;
    }
}

File: 1364-tuple-with-same-product/1364-tuple-with-same-product.cpp
class Solution {
    int nC2(int n) { return (n * (n - 1)) / 2; }

public:
    int tupleSameProduct(vector<int>& nums) {
        int n = nums.size();
        unordered_map<int, int> freq;
        for (int i = 0; i < n - 1; ++i)
            for (int j = i + 1; j < n; ++j)
                freq[nums[i] * nums[j]]++;

        int count = 0;
        for (auto i: freq)
        if(i.second>1)
            count += 8 * nC2(i.second);

        return count;
    }
};